### PR TITLE
DCD XDR pickle tests

### LINKED ISSUE #2878: Fail to pickle last frame of DCD, XDR files
## Expected behavior ##

The trajectory can be pickled no matter which frame it is current in.

## Actual behavior ##

XDR and DCD readers cannot pickle the last frame.
Raising the EOFError

## Code to reproduce the behavior ##

```python
u = mda.Universe(GRO_MEMPROT, XTC_MEMPROT)
u.trajectory[4] #  last frame
pickle.loads(pickle.dumps(u.trajectory))
EOFError: Trying to seek over max number of frames
```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) 2.0.0 dev
- Which version of Python (`python -V`)? 3.8
- Which operating system? Ubuntu 20

### PR DESCRIPTION
**Pull request by yuxuanzhuang**
Fixes #2878 

Changes made in this Pull Request:
 - add inclusive tests for pickle low-level dcd and xdr files
 - fix the bug that was introduced in #2723  
 - may further backport in #2908 
PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/lib/formats/libmdaxdr.pyx [hunk 1]
```diff
@@ -306,8 +306,14 @@ cdef class _XDRFile:
 
         # where was I
         current_frame = state[1]
-        self.seek(current_frame - 1)
-        self.current_frame = current_frame
+        if current_frame == self.offsets.size:
```
Comment (orbeckst) [inline package/MDAnalysis/lib/formats/libmdaxdr.pyx:309]: Switch the decision in the `if` statement to make the more common situation come first
```python
if current_frame < self.offsets.size:
     ...
elif current_frame == self.offsets.size:
    ...
else:
    #pragma: no cov
    raise RuntimeError("Invalid frame number {} > {} -- this should not happen.".format(current_frame, self.offsets.size)
```
and add defensive code.

Comment (orbeckst) [reply package/MDAnalysis/lib/formats/libmdaxdr.pyx:309]: Add comment under which circumstances we have to do funny things – namely when we need to serialize at the end of the trajectory.

**Comment by yuxuanzhuang** [comment]
I makes `DCDFile` to behave the same as `XDRFile`, i.e. not messing up with `seek`, but checking during `__setstate__`

**Comment by yuxuanzhuang** [comment]
The `pragma: no cover` is not honered in cython files (https://github.com/cython/cython/issues/3680)
Any reason we have to check for that here (aside from for the completeness), since it will be eventually checked in `seek`.

**Pull request merged**